### PR TITLE
feat(dataset) Finally remove dependencies on query.body

### PR DESCRIPTION
**Pull request by fpacifici**
This is the last PR of this stack:
#497 #499 #500 #504 

This finally removes all direct dependencies on Query body during query column processing.
Now that we depend exclusively on the Query data structure we can make deeper changes on how parsing and optimization happen since we have a single place to go to access the content of the query and the raw body is not leaking anywhere.

test plan
unit test (this is well covered, as it is visible from the amount of tests I had to fix)
ran sentry test and succeeded
the typechecker seems relatively happy.

### CODE REVIEW

**Comment by fpacifici** [comment]
I wanted to entirely remove the get_body method from Query as well but Request depends on it to build the request body that is used for logging. We should probably replace that method with a "serialize" or something like that that returns something hard to parse (like a string) for logging so none will be tempted of using it.

File: snuba/util.py [hunk 1]
```diff
@@ -238,24 +239,28 @@ def is_condition(cond_or_list):
     )
 
 
-def all_referenced_columns(query_body):
+def all_referenced_columns(query: Query):
```
Comment (fpacifici) [inline snuba/util.py]: I would like to move this as a public method of Query, but I cannot do that since it depends on a function that is in util and I cannot make Query depend on util (circular).
Taking a step back, the method this depends on is `columns_in_expr`. This is likely to become a method owned by an Expression class soon, so the problem will disappear

File: snuba/clickhouse/query.py [hunk 2]
```diff
@@ -14,36 +16,33 @@ class ClickhouseQuery:
 
     def __init__(self,
         dataset: Dataset,
-        # TODO: pass query only once util.column_expr will not depend on the full request
-        # body anymore.
-        request: Request,
+        query: Query,
+        settings: RequestSettings,
```
Comment (tkaemming) [inline snuba/clickhouse/query.py]: Starting to wonder if `RequestSettings` are named appropriately here…

File: snuba/util.py [hunk 3]
```diff
@@ -238,24 +239,28 @@ def is_condition(cond_or_list):
...
     """
     Return the set of all columns that are used by a query.
     """
-    col_exprs = []
+    col_exprs: Sequence[Any] = []
```
Comment (tkaemming) [inline snuba/util.py]:
```suggestion
    col_exprs: MutableSequence[Any] = []
```

**Review by tkaemming — APPROVED** [review]

File: snuba/clickhouse/query.py (hunk 2 above)
Comment (fpacifici) [reply snuba/clickhouse/query.py]: @manuzope you may want to look into how we are using them now.
